File: javascript/src/queue.ts
interface IQueue<T> {
  add(value: T): void;
  remove(): T | undefined;
  peek(): T | undefined;
  size(): number;
  toString(formatter: (value: T) => string): string;
}

export class Queue<T> implements IQueue<T> {
  private values: T[];

  constructor(arr?: T[]) {
    if (arr) {
      this.values = arr;
    } else {
      this.values = [];
    }
  }

  add(value: T): void {
    this.values.unshift(value);
  }

  remove(): T | undefined {
    return this.values.shift();
  }

  peek(): T | undefined {
    return this.values.at(0);
  }

  size(): number {
    return this.values.length;
  }

  toString(formatter: (value: T) => string): string {
    return "[" + this.values.map((v) => formatter(v)).join(",") + "]";
  }
}

File: java/src/main/java/com/example/Queue.java
package com.example;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import com.example.common.FormatterFunction;

interface IQueue<T> {
    public void add(T value);

    public Optional<T> remove();

    public Optional<T> peek();

    public int size();

    public String toString(FormatterFunction<T> formatter);
}

public class Queue<T> implements IQueue<T> {
    private ArrayList<T> values;

    public Queue() {
        this.values = new ArrayList<>();
    }

    public Queue(T[] arr) {
        this();
        for (int i = 0; i < arr.length; i++) {
            this.add(arr[i]);
        }
    }

    public Queue(List<T> values) {
        this.values = new ArrayList<>(values);
    }

    public void add(T value) {
        this.values.add(value);
    }

    public Optional<T> remove() {
        if (this.values.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(this.values.remove(this.values.size() - 1));
        }
    }

    public Optional<T> peek() {
        if (this.values.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(this.values.get(this.values.size() - 1));
        }
    }

    public int size() {
        return this.values.size();
    }

    @Override
    public String toString(FormatterFunction<T> formatter) {
        ArrayList<String> elems = new ArrayList<>();
        for (T t : this.values) {
            elems.add(formatter.apply(t));
        }
        return "[" + String.join(",", elems) + "]";
    }

}
